test/library/draft/DistributedMap/v3/ChainTable.chpl

```diff
@@ -1,391 +1,284 @@
+      this.valType = valType;
+      this.numBuckets = initialCapacity;
+      this.numEntries = 0;
     }
 
-    // a linked-list node which holds a hashtable entry
-    class ChainedTableEntry {
-        type keyType;
-        type valType;
-        var entry : rawTableEntry(keyType, valType);
-        var next : owned ChainedTableEntry(keyType, valType)?;
-
-        proc appendToTail(idx: uint) : uint {
-            if this.next != nil {
-                return this.next!.appendToTail(idx + 1);
-            } else {
-                this.next = new owned ChainedTableEntry(keyType, valType);
-                return idx;
-            }
-        }
+    // determine if the map has an entry for the given key
+    //  if so, return (true, bucket index, index within the bucket)
+    //  otherwise, return (false, bucket index, 0)
+    proc getFullSlotFor(key: keyType) : (bool, (uint, uint)) {
+        var bucket_idx = this._bucketIdx(key);
+        var (has_key, chain_idx) = this.buckets[bucket_idx].findIndexOf(key);
+        return (has_key, (bucket_idx, chain_idx));
     }
 
-    enum bucketInitState { uninit=0, initialized };
-
-    // a bucket for holding hashtable entries that all correspond the the same hash
-    //
-    // some small number of entries are stored in a local tuple, and the remainder
-    //  are stored on the heap in a linked list
-    //
-    // this type provides an interface for finding and modifying entries in the
-    //  bucket. Modifications are done by reference using the 'this()' accessor
-    record Bucket {
-        type keyType;
-        type valType;
-
-        // store some number of local entries to reduce # of cache misses when load factor is low
-        var localEntries: numLocalBucketSlots*rawTableEntry(keyType, valType);
+    // return the indices for a given key
+    //  if an entry is already present for the key, return (bucket index, its index within the bucket)
+    //  otherwise, return (bucket index, next open index in the bucket)
+    proc getSlotFor(key: keyType) : (uint, uint) {
+        var bucket_idx = this._bucketIdx(key);
+        var (has_key, chain_idx) = this.buckets[bucket_idx].findIndexOf(key);
+        return if has_key
+            then (bucket_idx, chain_idx)
+            else (bucket_idx, this.buckets[bucket_idx].firstAvailableIndex());
+    }
 
-        // store the remainder of the entries in a linked list on the heap
-        var chainHead: owned ChainedTableEntry(keyType, valType)?;
+    // copy a key-value pair into a slot with the given indices
+    proc fillSlot((bucket_idx, chain_idx): (uint, uint), in key: keyType, in val: valType) {
+        use MemMove;
 
-        var initState : bucketInitState = bucketInitState.uninit;
-        inline proc isInit() {
-            return this.initState == bucketInitState.initialized;
-        }
+        ref entry = this.buckets[bucket_idx][chain_idx];
+        if entry.status != entryStatus.full then this.numEntries += 1;
```

Comment:
Should the field this is incrementing be atomic?  It seems like there's some potential for races here

Reply:
I'm not sure whether it's strictly necessary since the calls to these methods are always wrapped in locks, but I'll make this atomic just in case.